debezium-embedded/src/main/java/io/debezium/embedded/EmbeddedEngine.java

```diff
@@ -342,6 +342,83 @@ public boolean hasError() {
         }
     }
 
+    /**
+     * Interfaces with the maybeFlush and commitRecord APIs to allow
+     * the ChangeConsumer to specify when commits happen
+     */
+    @ThreadSafe
+    public static interface RecordCommitter {
+        /**
+         * Marks a single record as processed, must be called for each
+         * record, and is threadsafe
+         * @param record the record to commit
+         */
+        void markProcessed(SourceRecord record) throws InterruptedException;
+
+        /**
+         * Marks a batch as finished, this may result in committing offsets/flushing
+         * data.
+         *
+         * Should be called when a batch of records is finished being processed.
+         *
+         * Is threadsafe
+         */
+        void markBatchFinished();
+
+    }
+
+    /**
+     * An interface that interacts between lists of SourceRecords, committing
+     * and the EventQueue
+     */
+    public static interface ChangeConsumer {
+
+        /**
+         * Handles a batch of records, calling the {@link RecordCommitter#markProcessed(SourceRecord)}
+         * for each record and {@link RecordCommitter#markBatchFinished()} when this batch is finished
+         *
+         * @param committer the committer that indicates to the system that we are finished
+         * @param records the records to be processed
+         */
+        void handleBatch(RecordCommitter committer, List<SourceRecord> records) throws Exception;
+    }
+
+    private static ChangeConsumer buildDefaultChangeConsumer(Consumer<SourceRecord> consumer) {
+        return new ChangeConsumer() {
+            /**
+             * the default implementation that is compatible with the old Consumer api.
+             *
+             * On every record, it calls the consumer, and then only marks the record
+             * as processed when accept returns, additionally, it handles StopConnectorExceptions
+             * and ensures that we all ways try and mark a batch as finished, even with exceptions
+             *
+             * @param committer the committer that indicates to the system that we are finished
+             * @param records the records to be processed
+             * @throws Exception
+             */
+            @Override
+            public void handleBatch(RecordCommitter committer, List<SourceRecord> records) throws Exception {
+                try {
+                    for (SourceRecord record : records) {
+                        try {
+                            consumer.accept(record);
+                            committer.markProcessed(record);
+                        }
+                        catch (StopConnectorException ex) {
```

Comment:
This exception seems a bit bogus, it's actually never raised other than in test code. But that's something to address separately, I reckon.